gnovm/stdlibs/std/params.go

```diff
@@ -8,65 +8,127 @@ import (
 	gno "github.com/gnolang/gno/gnovm/pkg/gnolang"
 )
 
-// ParamsInterface is the interface through which Gno is capable of accessing
+// ParamsSetterInterface is the interface through which Gno is capable of accessing
 // the blockchain's params.
 //
 // The name is what it is to avoid a collision with Gno's Params, when
 // transpiling.
+
 type ParamsInterface interface {
-	SetString(key, val string)
-	SetBool(key string, val bool)
-	SetInt64(key string, val int64)
-	SetUint64(key string, val uint64)
-	SetBytes(key string, val []byte)
+	SetString(key ParamKey, val string)
+	SetBool(key ParamKey, val bool)
+	SetInt64(key ParamKey, val int64)
+	SetUint64(key ParamKey, val uint64)
+	SetBytes(key ParamKey, val []byte)
+}
+
+type ParamKey struct {
+	Realm  string
+	Prefix string
+	Key    string
+	Type   string
+}
+
+func NewParamKey(m *gno.Machine, prefix, key string, kind string) (ParamKey, error) {
+	// validate key.
+	if err := validate(prefix, key, kind); err != nil {
+		es := err.Error()
+		m.Panic(typedString(es))
+		return ParamKey{}, err
+	}
+	_, realm := currentRealm(m)
+
+	return ParamKey{
+		Realm:  realm,
+		Prefix: prefix,
+		Key:    key,
+		Type:   kind,
+	}, nil
+}
+
+// String representation of ParamKey in the format:
+// <module>:(<realm>".")?<key>
+func (pk ParamKey) String() string {
+	pks := ""
+	if pk.Prefix == "" {
+		pks = fmt.Sprintf("vm:%s.%s", pk.Realm, pk.Key)
```

Comment:
it should be fmt.Sprintf("%s:%s.%s", pk.Module, pk.Realm, pk.Key). 
"vm:" should not be hard coded here so different modules can have "realms" too.

type ParamKey struct {
	Module string // NOTE comes first
	Realm  string
	Key    string
	Type   string
}

Reply:
If 'auth:gno.land/r/myreal.foo.string' is considered a valid key, how can we differentiate between a user's intention to modify a predefined module parameter (such as 'auth:max_memo_bytes' or 'auth:tx_size_cost_per_byte') and their intention to add and set an arbitrary parameter 'gno.land/r/myreal.foo.string', which is not a valid key in the auth module?

Reply:
no, those are exceptions.
And if they come from r/sys/params, it is considered valid; r/sys/params must find a way to authenticate, or fix the implementation so it is safe.